### DISCUSSION TITLE
Has anyone got termux x11 apps to work?

### DISCUSSION THREAD

**Discussion by Quackdoc** [slint-ui/slint#7228] [Q&A] [answered]
Im trying to use slint in termux, and while I kinda managed to get it working over wayland, I have found it to be very buggy. I was hoping to get it working over x11 but winit here seems to fail for some reason.

When compiling the application I use `cargo build --release --target aarch64-unknown-linux-gnu --no-default-features --features "winit_software"` however I have also tried femtovg with the same issues.

Compiling with gnu is necessary as termux has no musl environment and when you compile with bionic a lot of rust software wrongfully assume that it will be running in the "android ecosystem" and disable support for x11 and wayland, and make some other breaking assumptions.

To launch with termux-x11 you can run the below command, in this case I am trying a software called krokiet which is part of czkawka

I think maybe winit doesn't support dlopen for x11? if so that is probably the cause of these issues but I haven't had time to investigate it yet.

```
➜  ~ termux-x11 :1 -xstartup "XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=1 grun ./krokiet"
22:32:27.505 [INFO] czkawka_core::common: App version: 8.0.0, release mode, rust 1.83.0 (2024-11-26), os Linux Unknown [aarch64 64-bit], 5 cpu/threads, features(0): []
thread 'main' panicked at krokiet/src/main.rs:69:33:
Failed to create main window: Other("Could not initialize backend.\nError from Winit backend: Error initializing winit event loop: the requested operation is not supported by Winit\nNo backends configured.")
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::result::unwrap_failed
   3: krokiet::main
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
3250 exited, status=101
```
You can also try using ssh -X and the results are the same. Below is the similar command but trying over `ssh -X` and with RUST_BACKTRACE=full

```
➜  ~ XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=full grun ./krokiet
22:37:18.964 [INFO] czkawka_core::common: App version: 8.0.0, release mode, rust 1.83.0 (2024-11-26), os Linux Unknown [aarch64 64-bit], 6 cpu/threads, features(0): []
thread 'main' panicked at krokiet/src/main.rs:69:33:
Failed to create main window: Other("Could not initialize backend.\nError from Winit backend: Error initializing winit event loop: the requested operation is not supported by Winit\nNo backends configured.")
stack backtrace:
   0:       0x7ee75e0cc0 - <std::sys::backtrace::BacktraceLock::print::DisplayBacktrace as core::fmt::Display>::fmt::ha22ce564a7afbb71
   1:       0x7ee760c838 - core::fmt::write::h7f405d3ca1a38e64
   2:       0x7ee75dbc20 - std::io::Write::write_fmt::hed8e58bb4069d8a1
   3:       0x7ee75e0b74 - std::sys::backtrace::BacktraceLock::print::h8ad1c1f9fe1e5396
   4:       0x7ee75e1c64 - std::panicking::default_hook::{{closure}}::h5a3e35b94e8985bf
   5:       0x7ee75e1aa8 - std::panicking::default_hook::h84f6fe1e5c50e46e
   6:       0x7ee75e2298 - std::panicking::rust_panic_with_hook::hcc55850fbc210339
   7:       0x7ee75e20ec - std::panicking::begin_panic_handler::{{closure}}::hd8680ae1abf66a51
   8:       0x7ee75e11bc - std::sys::backtrace::__rust_end_short_backtrace::h1c3166319815e2de
   9:       0x7ee75e1dac - rust_begin_unwind
  10:       0x7ee6157798 - core::panicking::panic_fmt::h721dbb42e0280a17
  11:       0x7ee6157afc - core::result::unwrap_failed::h79836a03ae9dd335
  12:       0x7ee643108c - krokiet::main::hc92e83b62d6b6feb
  13:       0x7ee63d67c0 - std::sys::backtrace::__rust_begin_short_backtrace::hdc28b62fa2059d5b
  14:       0x7ee63f0da4 - std::rt::lang_start::{{closure}}::ha3b93459ea8c6f46
  15:       0x7ee75d3a3c - std::rt::lang_start_internal::h4d9c6a98427dc95d
  16:       0x7ee643156c - main
  17:       0x7ee5c99c80 - __libc_start_call_main
  18:       0x7ee5c99d60 - __libc_start_main_alias_2
  19:       0x7ee6158170 - _start
  20:                0x0 - <unknown>
```

EDIT: x11-dl should use dlopen, but im not sure if winit has other dependancies that may break this

**Comment by Quackdoc** [comment, marked as answer]
I actually solved this myself. It turns out that winit does not report missing deps properly unless I had run a winit example. How I got czkawka/krokiet working

```sh
cargo build --release --target aarch64-unknown-linux-gnu
#copy binary to android
ssh -Y -C -p 8022 u0_a207@192.168.1.201  #needs -Y and -C helps a lot over wifi
apt install ttf-dejavu-glibc libxcursor-glibc fontconfig-utils-glibc libxkbcommon-glibc libxi-glibc libx11-glibc grun #it's important to install all of these as slint winit/slint will fail without reporting a proper missing dep
chmod +x ./krokiet
XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=1 grun ./krokiet
```


If you plan on using termux-x11 you cannot start krokiet with xstartup (nor would you want to) it is reccomended to use a window manager such as dwm or i3 to make sure you get the full usable space of you phone, then run `XDG_DATA_DIRS=$PREFIX/glibc/share SLINT_DEFAULT_FONT=$XDG_DATA_DIRS/fonts  RUST_BACKTRACE=1 DISPLAY=:1 grun ./krokiet`

**Comment by ghost** [comment]
Termux repository has proot-distro, which can install archlinux ARM version for termux. Can slint compile in archlinux, which is installed by termux  proot-distro? It isn't a distribution directly installed on bare metal, but installed in termux proot-distro command.
